[FEATURE] Introduce interactive viewer plugins.

Milotrince opened the pull request:
## Description
nice-to-have feature, allows users to integrate interactive viewer plugins.

- Remove dependency on `pynput`
- `gs.vis.keybindings.Keybind` callback registration
- in `gs.options.ViewerOptions`:
  - `disable_keyboard_shortcuts` replaced with `disable_default_keybinds`
  - add `disable_help_text`

## Related Issue

Resolves https://github.com/Genesis-Embodied-AI/Genesis/issues/713
Resolves https://github.com/Genesis-Embodied-AI/Genesis/issues/1556

## Motivation and Context
- `ViewerPlugin` enables directly interfacing with pyviewer mouse/keyboard controls.
- Replaces `pynput`
- Simple keybind callback registration

## How Has This Been / Can This Be Tested?
added `tests/test_viewer.py`

## Screenshots (if appropriate):

<img width="453" height="370" alt="Screenshot 2026-01-30 at 16 08 08" src="https://github.com/user-attachments/assets/8885f534-441b-486a-837f-5855b1aab00e" />
registered keys automatically show in help text :)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Milotrince commented:
- Should we keep under `genesis/ext/pyrender/interaction/plugins/` or `genesis/vis/plugins` ?
- Do we want to use Vec3, Quat, Color, Ray, etc. classes in the codebase?
- now that pyrender viewer mouse/keyboard interaction is exposed, we can get rid of `pynput` in examples

duburcqa commented:
@Milotrince If you could finish this PR I would be happy to review it and merge it :)

Milotrince commented:
@duburcqa Could I get your opinion on the state of the PR? there are some design decisions I'd like your input on!
- removed pynput, see examples/* changes for new usage
- I kept the mouse spring interaction implementation and moved it as a plugin
- there existed a register key system in pyrender viewer but was not exposed in genesis viewer. I made a Keybind/Keybindings class and exposed register_keybind to viewer

I still need to add tests.

duburcqa commented:
> Could I get your opinion on the state of the PR? there are some design decisions I'd like your input on!

Sure ! I will do it this week ! Could you fix the conflicts in the meantime? Thank you :)

duburcqa commented on examples/IPC_Solver/ipc_arm_cloth.py:
```diff
@@ -248,99 +262,57 @@ def reset_scene():
-    stop = False
     step_count = 0
 
-    while not stop:
```
Why the stop mechanism had to be removed in favor of try-except?

duburcqa commented on examples/sensors/lidar_teleop.py:
```diff
@@ -193,48 +136,79 @@ def apply_pose_to_all_envs(pos_np: np.ndarray, quat_np: np.ndarray):
+        target_euler[2] += KEY_DANGLE
+
+    def yaw_cw():
+        target_euler[2] -= KEY_DANGLE
```
This is very verbose for little benefit. What about only exposing `translate`, `rotate`, with some "axis: int, is_negative: bool" argument or just `axis: np.ndarray`, and use `partial` to keybind them? That would be more flexible and simpler I think.

Milotrince commented on examples/drone/interactive_drone.py:
```diff
@@ -145,14 +75,41 @@ def main():
     scene.build()
 
+    # Register keybindings
+    from pyglet.window import key
```
leaving a note here: on mac (untested with other machines), when i `import pyglet` or `pyglet.window.key` at the top level before viewer does, I get
```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSApplication macOSVersion]: unrecognized selector sent to instance 0x1618b7df0'

*** First throw call stack:
(
        0   CoreFoundation                      0x0000000198f498ec __exceptionPreprocess + 176
        1   libobjc.A.dylib                     0x0000000198a22418 objc_exception_throw + 88
        2   CoreFoundation                      0x000000019902b7e8 +[NSObject(NSObject) instanceMethodSignatureForSelector:] + 0
...
        87  python3.12                          0x000000010436e1f0 builtin_exec + 404
        88  python3.12                          0x00000001042bb45c cfunction_vectorcall_FASTCALL_KEYWORDS + 92
        89  python3.12                     
libc++abi: terminating due to uncaught exception of type NSException
```
By claude's suggestion I added it as a lazily loading module to safely expose  `key` at `genesis.vis.keybindings`.

Milotrince commented on examples/drone/interactive_drone.py:
```diff
@@ -1,112 +1,41 @@
+        """All rotors spin faster"""
+        self.thrust += self.thrust_delta
+        self.rpms = [self.thrust] * 4
+        self.rpms = np.clip(self.rpms, 0, 25000)
```
the use of thrust and rpms is how it was before. I'll update it based on your suggestions!

Milotrince commented on genesis/engine/sensors/raycaster.py:
```diff
@@ -195,8 +85,7 @@ def kernel_cast_rays(
         ray_direction_world = ti_normalize(ti_transform_by_quat(ray_dir_local, link_quat), gs.EPS)
 
-        # --- 2. BVH Traversal ---
-        # FIXME: this duplicates the logic in LBVH.query() which also does traversal
```
I removed it because I think the FIXME note isn't quite accurate, `raycaster.py` does ray-AABB tests while `LBVH.query()` checks AABB-AABB. There might be a bit of duplicated logic but i think it'd be more complicated to split it up into smaller `ti.func`s so may as well leave as is.

duburcqa commented:
@Milotrince CI is failing!

Milotrince commented:
updated snapshot pr: https://huggingface.co/datasets/Genesis-Intelligence/snapshots/discussions/7

Milotrince commented on genesis/vis/viewer.py:
```diff
@@ -40,16 +41,18 @@ def __init__(self, options: "ViewerOptions", context):
+        if not options.disable_default_keybinds:
+            from genesis.vis.viewer_plugins import DefaultControlsPlugin
+
+            self._viewer_plugins.append(DefaultControlsPlugin())
```
it's not necessary, moved it up!

duburcqa commented on .github/workflows/examples.yml:
```diff
@@ -54,7 +54,6 @@ jobs:
         run: |
           python -m pip install --upgrade pip setuptools wheel
           pip install torch --index-url https://download.pytorch.org/whl/cpu
-          pip install -e '.[dev,usd]' pynput
```
`pip install -e '.[dev,usd]'`
